File: spring/core/SpringBeanLifecycles/src/main/java/demos/lifecycle/lazy/creation/SampleShop.java
package demos.lifecycle.lazy.creation;

import demos.lifecycle.common.PaymentEngine;
import demos.lifecycle.common.PricingEngine;
import demos.lifecycle.common.Shop;
import demos.lifecycle.common.StockCheckEngine;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;

public class SampleShop implements Shop {
    private PricingEngine pricingEngine;
    private StockCheckEngine stockCheckEngine;
    private PaymentEngine paymentEngine;

    @Autowired
    public void setPricingEngine(PricingEngine pricingEngine) {
        this.pricingEngine = pricingEngine;
    }

    @Autowired
    public void setStockCheckEngine(StockCheckEngine stockCheckEngine) {
        this.stockCheckEngine = stockCheckEngine;
    }

    @Autowired
    public void setPaymentEngine(PaymentEngine paymentEngine) {
        this.paymentEngine = paymentEngine;
    }

    @Override
    public boolean makePurchase(String itemNo, int quantity, String cardNo) {
        if (stockCheckEngine.check(itemNo) >= quantity) {
            double charge = pricingEngine.price(itemNo, quantity);
            if (paymentEngine.authorize(cardNo, charge)) {
                return true;
            }
        }
        return false;
    }

    @PostConstruct
    public void begin() {
        System.out.println("Spring context just created a shop");
    }

    @PreDestroy
    public void end() {
        System.out.println("Spring context just released a shop");
    }
}

File: spring/core/AnnotationDrivenDIWithKotlin/src/main/kotlin/demos/spring/notes/v10/SampleShop.kt
package demos.spring.notes.v10

import demos.spring.notes.common.PaymentEngine
import demos.spring.notes.common.PricingEngine
import demos.spring.notes.common.Shop
import demos.spring.notes.common.StockCheckEngine
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service("shop")
class SampleShop @Autowired constructor(
        @Qualifier("pricing") private val pricingEngine: PricingEngine,
        @Qualifier("stock") private val stockCheckEngine: StockCheckEngine,
        @Qualifier("payment") private val paymentEngine: PaymentEngine) : Shop {

    var shopName: String = ""
        @Value("#{dataSource.CEO.toUpperCase() == 'STEVE JOBS' ? 'Apple' : 'Microsoft'}")
        set(value) {
            field = value
        }

    var openDuringWeekends = false
        @Value("#{dataSource.openDays matches '.*(Saturday|Sunday).*'}")
        set(value) {
            field = value
        }


    override fun makePurchase(itemNo: String, quantity: Int, cardNo: String): Boolean {
        if (stockCheckEngine.check(itemNo) >= quantity) {
            val charge = pricingEngine.price(itemNo, quantity)
            if (paymentEngine.authorize(cardNo, charge)) {
                return true
            }
        }
        return false
    }
}
